realsense2_camera/src/camera_driver.cpp

```diff
@@ -0,0 +1,912 @@
+        }
+        else if (frame.second.is<rs2::depth_frame>() && _clipping_distance > 0) 
+        {
+            clip_depth(frame.second, _clipping_distance);
+        }
+    }
+}
+
+CameraDriver::CameraDriver(ros::NodeHandle nh, ros::NodeHandle pnh, const std::string& serial_no) :
+    _nh(nh),
+    _pnh(pnh),
+    _serial_no(serial_no)
+{
+}
+
+std::string CameraDriver::getSerialNo() const
+{
+    return _serial_no;
+}
+
+void CameraDriver::reset()
+{
+    _camera = {};
+    if (_device) 
+    {
+        try 
+        {
+            _device.hardware_reset();
+        }
+        catch(const std::exception& ex) 
+        {
+            ROS_ERROR("An exception has been thrown trying reset device <%s>: %s", _serial_no.c_str(), ex.what());
+        }
+
+        _device = rs2::device();
+    }
+}
+
+void CameraDriver::setCallback(FrameCallback callback)
+{
+    _frame_callback = callback;
+}
+
+FramesPtr CameraDriver::getFrames(double timeout, ros::Time stamp)
+{
+    std::lock_guard<std::mutex> lock(_get_frames_mutex);
+
+    if (!_camera)
+    {
+        return  {};
+    }
+
+    auto raw_frames = std::dynamic_pointer_cast<PolledRealsenseNode>(_camera)->getFrames(timeout * 1000, stamp);
+    if (!raw_frames)
+    {
+        return {};
+    }
+
+    return processFrames(*raw_frames);
+}
+
+FramesPtr CameraDriver::getLatestFrames()
+{
+    std::lock_guard<std::mutex> lock(_get_frames_mutex);
+    if (_latest_processed_frames)
+    {
+        return _latest_processed_frames;
+    }
+    else if (!_latest_frames)
+    {
+        return {};
+    }
+
+    _latest_processed_frames = processFrames(*_latest_frames);
+    return _latest_processed_frames;
+}
+
+ros::Time CameraDriver::getLatestStamp()
+{
+    std::lock_guard<std::mutex> lock(_get_frames_mutex);
+
+    if (!_camera) 
+    {
+        return ros::Time(0);
+    }
+
+    if (!_latest_frames || _latest_frames->empty())
+    {
+        return ros::Time(0);
+    }
+
+    return ros::Time(std::dynamic_pointer_cast<PolledRealsenseNode>(_camera)->getStamp(_latest_frames->begin()->second));
+}
+
+void CameraDriver::setFrameBuffer(FramesPtr frame_buffer)
+{
+    _frame_buffer = frame_buffer;
+}
+
+void CameraDriver::start()
+{
+    if (!_camera)
+    {
+        return;
+    }
+    std::dynamic_pointer_cast<PolledRealsenseNode>(_camera)->startStreams();
+}
+
+void CameraDriver::stop()
+{
+    if (!_camera)
+    {
+        return;
+    }
+    std::dynamic_pointer_cast<PolledRealsenseNode>(_camera)->stopStreams();
+}
+
+double CameraDriver::getProjectorTemperature() const
+{
+    if (!_camera)
+    {
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+
+    // https://github.com/IntelRealSense/librealsense/issues/866#issuecomment-357461253
+    auto dbg = _device.as<rs2::debug_protocol>();
+    static const std::vector<uint8_t> cmd = { 0x14, 0, 0xab, 0xcd, 0x2a, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
```

Comment:
Move this to the header and give a more descriptive name for what it's doing, e.g. GET_PROJECTOR_TEMP_CMD